opentelemetry-sdk/src/export/logs/mod.rs

```diff
@@ -18,9 +18,18 @@ use std::fmt::Debug;
 ///
 #[derive(Debug)]
 pub struct LogBatch<'a> {
-    /// The data field contains a slice of tuples, where each tuple consists of a reference to
-    /// a `LogRecord` and a reference to an `InstrumentationScope`.
-    data: &'a [(&'a LogRecord, &'a InstrumentationScope)],
+    data: LogBatchData<'a>,
+}
+
+/// The `LogBatchData` enum represents the data field of a `LogBatch`.
+/// It can either be:
+/// - A shared reference to a vector of tuples, where each tuple consists of a `LogRecord` and an `InstrumentationScope`.
+/// - Or it can be a slice of tuples, where each tuple consists of a reference to a `LogRecord` and a reference to an `InstrumentationScope`.
+#[derive(Debug)]
+#[allow(clippy::vec_box)] // Clippy complains about using Box in a Vec, but it's done here for performant moves of the data between channel and the vec.
```

Comment:
Clippy complains about using [`Box` inside a `vec`](https://rust-lang.github.io/rust-clippy/master/index.html#/vec_box). Since `vec` allocates its elements on heap, boxing introduces unnecessary redirection.

The boxing happens in [`LogProcessor::emit()`](https://github.com/open-telemetry/opentelemetry-rust/blob/9a8ad95ee8a9f777427ad3742b4e10fd6acc298c/opentelemetry-sdk/src/logs/log_processor.rs#L281-L286). I think we do this for faster movement of the data from the channel to the vec. We may need to revisit this. It's outside of the scope of this PR.